src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetService.cs

```diff
@@ -0,0 +1,160 @@
+        if (!await contentDbContext.ReleaseFiles
+                .Where(rf => rf.ReleaseVersionId == releaseFile.ReleaseVersionId)
+                .Where(rf => rf.File.SubjectId == releaseFile.File.SubjectId)
+                .Where(rf => rf.File.Type == FileType.Metadata)
+                .AnyAsync(cancellationToken: cancellationToken))
+        {
+            errors.Add(CreateError(
+                code: ValidationMessages.NoMetadataFile.Code,
+                message: ValidationMessages.NoMetadataFile.Message,
+                releaseFileId: releaseFile.Id));
+        }
+
+        return errors.Count == 0 ? Unit.Instance : ValidationUtils.ValidationResult(errors);
+    }
+
+    private async Task<DataSet> CreateDataSet(
+        ReleaseFile releaseFile,
+        CancellationToken cancellationToken)
+    {
+        var dataSet = new DataSet
+        {
+            Status = DataSetStatus.Draft,
+            Title = releaseFile.Name!,
+            Summary = releaseFile.Summary ?? "",
+            PublicationId = releaseFile.ReleaseVersion.PublicationId
+        };
+
+        publicDataDbContext.Add(dataSet);
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+
+        return dataSet;
+    }
+
+    private async Task<DataSetVersion> CreateDataSetVersion(
+        DataSet dataSet,
+        ReleaseFile releaseFile,
+        Guid instanceId,
+        CancellationToken cancellationToken)
+    {
+        var dataSetVersion = new DataSetVersion
+        {
+            ReleaseFileId = releaseFile.Id,
+            DataSetId = dataSet.Id,
+            Status = DataSetVersionStatus.Processing,
+            Imports =
+            [
+                new DataSetVersionImport
+                {
+                    InstanceId = instanceId, Stage = DataSetVersionImportStage.Pending
+                }
+            ],
+            Notes = "",
+            VersionMajor = dataSet.LatestLiveVersion?.VersionMajor ?? 1,
+            VersionMinor = dataSet.LatestLiveVersion?.VersionMinor + 1 ?? 0
+        };
+
+        dataSet.Versions.Add(dataSetVersion);
+        dataSet.LatestDraftVersion = dataSetVersion;
+
+        publicDataDbContext.DataSets.Update(dataSet);
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+
+        return dataSetVersion;
+    }
+
+    private async Task UpdateFilePublicDataSetVersionId(
+        ReleaseFile releaseFile,
+        Guid dataSetVersionId,
+        CancellationToken cancellationToken)
+    {
+        releaseFile.File.PublicDataSetVersionId = dataSetVersionId;
+        await contentDbContext.SaveChangesAsync(cancellationToken);
+    }
+
+    private static ErrorViewModel CreateError(
+        string code,
+        string message,
+        Guid releaseFileId)
+    {
+        return new ErrorViewModel
+        {
+            Code = code,
+            Message = message,
+            Path = nameof(releaseFileId),
```

Comment:
This probably isn't the best thing to use for the name. Could use the request model itself instead i.e.

```cs
Path = nameof(InitialDataSetVersionCreateRequest.ReleaseFileId).ToLowerFirst()
```